### PR TITLE
intel_adsp/ace: power: Lock interruption when power gate fails

### PR DESCRIPTION
**Pull request by ceolin**
In case the core is not power gated, waiti will restore intlevel. In this case we lock interruption after it.

### CODE REVIEW

**Review by dcpleung — DISMISSED** [review]

File: soc/xtensa/intel_adsp/ace/power.c [hunk 1]
```diff
@@ -176,6 +176,11 @@ void power_gate_entry(uint32_t core_id)
 	soc_cpus_active[core_id] = false;
 	sys_cache_data_flush_range(soc_cpus_active, sizeof(soc_cpus_active));
 	k_cpu_idle();
+
+	/* It is unlikely we get in here, but when this happens
+	 * we need to lock interruptions again.
+	 */
+	(void)irq_lock();
```
Comment (andyross) [inline soc/xtensa/intel_adsp/ace/power.c:183]: Needs to be arch_irq_lock(), or else an RSIL instruction or some other sequence to directly set PS.  When SMP=y, irq_lock() is a global recursive lock that emulates the traditional API for the benefit of drivers not yet ported to spinlocks, not what you want at all (and dangerous if there are other irq_lock() users as it looks to me like this call is unmached and would produce a deadlock if someone else tries it).

File: soc/xtensa/intel_adsp/ace/power.c [hunk 2]
```diff
@@ -176,6 +176,11 @@ void power_gate_entry(uint32_t core_id)
...
 	z_xt_ints_off(0xffffffff);
```
Comment (andyross) [inline soc/xtensa/intel_adsp/ace/power.c:184]: Looks like this line is already masking all interrupts, though at one level higher in the abstraction stack (INTENABLE instead of PS.INTLVL).  So why do you need both?

**Review by andyross — DISMISSED** [review]
Needs to use the correct API to mask interrupts.   But I'm confused as to why this is needed or what the circumstances are under which the WAITI returns?  If this is a "should never happen" maybe a k_panic() is a better idea?

File: soc/xtensa/intel_adsp/ace/power.c (hunk 2 above)
Comment (ceolin) [reply soc/xtensa/intel_adsp/ace/power.c:184]: Because Zephyr only looks PS.INTENABLE and assumes that interruptions are not masked and triggers an assert.

File: soc/xtensa/intel_adsp/ace/power.c (hunk 1 above)
Comment (ceolin) [reply soc/xtensa/intel_adsp/ace/power.c:183]: Ouch forgot about and thought it was just an indirection to `arch_irq_lock`. Yep, it has to be `arch_irq_lock()`, It is rstored later calling directly `rsil`.

**Comment by ceolin** [comment]
> Needs to use the correct API to mask interrupts. But I'm confused as to why this is needed or what the circumstances are under which the WAITI returns? If this is a "should never happen" maybe a k_panic() is a better idea?

When this can happen in HW is not clear for me, I had removed it in a different pr and was asked to get it back saying that something could happen and prevent the power gate. That's said, the simulator that we have does not power gate the core and this is need as well.

**Comment by ceolin** [comment]
@andyross would mind check it again ?

**Review by andyross — COMMENTED** [review]
> When this can happen in HW is not clear for me, I had removed it in a different pr and was asked to get it back saying that something could happen and prevent the power gate.

Can you get a proper explanation from whoever made that request? :)

Again, as I read this patch it's a noop.  The adjacent line is already masking all interrupts through a different mechanism.  Both/either have the effect of causing the core to resume handling thread code but with interrupts masked, which can't possibly be correct: you wouldn't be able to receive the IDC interrupt delivering the "shut down now" command from the host (exactly the action that seems to have failed in the first place).  So the only way to recover would be to hard-bounce the whole DSP anyway, which obviously doesn't care about interrupt masking state.

Basically I just don't see it.  If this needs to be handled, this isn't the right way to handle it.  If it shouldn't ever happen but might, it should be a panic.

All that said: this isn't a -1 over this issue.  It's a noop, after all, and it's not my job to tell Intel how to manage its own hardware.  (For propriety reasons.  But also to keep powder dry for fights over SOF where I do have more of a stake, heh.)

**Comment by ceolin** [comment]
> > When this can happen in HW is not clear for me, I had removed it in a different pr and was asked to get it back saying that something could happen and prevent the power gate.
> 
> Can you get a proper explanation from whoever made that request? :)
> 
> Again, as I read this patch it's a noop. The adjacent line is already masking all interrupts through a different mechanism. Both/either have the effect of causing the core to resume handling thread code but with interrupts masked, which can't possibly be correct: you wouldn't be able to receive the IDC interrupt delivering the "shut down now" command from the host (exactly the action that seems to have failed in the first place). So the only way to recover would be to hard-bounce the whole DSP anyway, which obviously doesn't care about interrupt masking state.

Nope, if that code is executed, we will return to the idle thread that will end up calling `pm_state_exit_post_ops` that restores interruptions properly.

> 
> Basically I just don't see it. If this needs to be handled, this isn't the right way to handle it. If it shouldn't ever happen but might, it should be a panic.

I know we are talking upstream here, but we have simulator used for validation where this happens (the soc is not power gated) and as it was told me that this may happen in HW, it seems to me that it is better to have it, also the system gets idle and execution continues correctly.

> 
> All that said: this isn't a -1 over this issue. It's a noop, after all, and it's not my job to tell Intel how to manage its own hardware. (For propriety reasons. But also to keep powder dry for fights over SOF where I do have more of a stake, heh.)

That is not noop, Zephyr only sees PS.INTLEVEL to check if interruptions are locked, without it, e.g any use of `_current_cpu` will trigger an assert.

**Review by kv2019i — DISMISSED** [review]

**Review by dcpleung — DISMISSED** [review]

**Comment by andyross** [comment]
Not a -1, but still asking nicely if the commit message can be clarified to explain why this is necessary and under what circumstances this code can execute.  Because it's not actually testable, right?  Has this actually been observed to happen?

**Review by kv2019i — APPROVED** [review]
Very minor typo in commit "thatthe", but good otherwise. Thanks @nashif !

**Review by lyakh — APPROVED** [review]

**Review by tmleman — APPROVED** [review]

**Review by nashif — APPROVED** [review]

**Comment by nashif** [comment]
> Very minor typo in commit "thatthe", but good otherwise. Thanks @nashif !

fixed

**Comment by nashif** [comment]
@andyross  can you please take another look at this?
